File: Yandex/YandexCup/2022/backend/quals/A.cpp

#include <iostream>
#include <set>
#include <vector>
#include <map>
#include <string>

using namespace::std;
#define int long long

void rm_old(map<int, set<int>> &val, int old, int num)
{
    auto old_it = val.find(old);
    if (old_it != val.end()) {
        old_it->second.erase(num);
        if (old_it->second.empty()) {
            val.erase(old_it);
        }
    }
}

signed main()
{
    int n, m, q;

    cin >> n >> m >> q;

    vector<set<int>> data(n);
    vector<int> resets(n);
    map<int, set<int>> values;
    vector<int> scores(n, 0);
    for (int i=0; i < n; ++i) {
        values[0].insert(i);
    }

    for (int i=0; i < q; ++i) {
        string cmd;
        cin >> cmd;
        if (cmd == "GETMAX") {
                cout << *(values.rbegin()->second.begin()) + 1 << endl;
        } else if (cmd == "GETMIN") {
                cout << *(values.begin()->second.begin()) + 1 << endl;
        } else if (cmd == "RESET") {
            int num;
            cin >> num;
            --num;
            int old_score = scores.at(num);
            rm_old(values, old_score, num);
            resets.at(num) += 1;
            data.at(num).clear();
            int new_score = resets.at(num) * m;
            values[new_score].insert(num);
            scores.at(num) = new_score;
        } else {
            int num, serv;
            cin >> num >> serv;
            --num;
            --serv;
            int old_score = scores.at(num);
            rm_old(values, old_score, num);
            data.at(num).insert(serv);
            int new_score = resets.at(num) * (m - data.at(num).size());
            values[new_score].insert(num);
            scores.at(num) = new_score;
        }
    }
}

File: Yandex/YandexCup/2022/backend/quals/A.py

def inp():
    return list(map(int, input().split()))

n, m, q = inp()
data = [set() for _ in range(n)]
resets = [0] * n

max_ra = 0
max_num = 0
min_ra = 0
min_num = 0

for _ in range(q):
    a = input()
    if a == 'GETMAX':
        print(max_num + 1)
        continue
    elif a == 'GETMIN':
        print(min_num + 1)
        continue
    elif a.startswith('RESET'):
        _, num = a.split()
        num = int(num) - 1
        resets[num] += 1
        data[num] = set()
        new_val = resets[num] * m
    else:
        _, dc, s = a.split()
        dc = int(dc) - 1
        s = int(s) - 1
        data[dc].add(s)
        new_val = resets[dc] * (m - len(data[dc]))
        num = dc

    if (new_val == max_ra):
        max_num = min(max_num, num)
    elif (new_val > max_ra):
        max_ra = new_val
        max_num = num
    if (new_val == min_ra):
        min_num = min(min_num, num)
    elif (new_val < min_ra):
        min_ra = new_val
        min_num = num

